builder/dockerfile/copy.go

```diff
@@ -288,33 +289,62 @@ func walkSource(source builder.Source, origPath string) ([]string, error) {
 	return subfiles, nil
 }
 
-type sourceDownloader func(string) (builder.Source, string, error)
+type sourceDownloader func(url, dest string) (builder.Source, string, error)
 
 func newRemoteSourceDownloader(output, stdout io.Writer) sourceDownloader {
-	return func(url string) (builder.Source, string, error) {
-		return downloadSource(output, stdout, url)
+	return func(url, dest string) (builder.Source, string, error) {
+		return downloadSource(output, stdout, url, dest)
 	}
 }
 
-func errOnSourceDownload(_ string) (builder.Source, string, error) {
+func errOnSourceDownload(_, _ string) (builder.Source, string, error) {
 	return nil, "", errors.New("source can't be a URL for COPY")
 }
 
-func downloadSource(output io.Writer, stdout io.Writer, srcURL string) (remote builder.Source, p string, err error) {
+func getTargetFilename(path string) string {
+	if path != "" && !strings.HasSuffix(path, "/") {
+		if filename := filepath.Base(filepath.FromSlash(path)); filename != "" {
+			return filename
+		}
+	}
+	return ""
+}
+
+func getTargetFilenameFromResp(resp *http.Response) string {
+	// Guess filename based on Content-Disposition
+	if contentDisposition := resp.Header.Get("Content-Disposition"); contentDisposition != "" {
+		if _, params, err := mime.ParseMediaType(contentDisposition); err == nil {
+			if params["filename"] != "" && !strings.HasSuffix(params["filename"], "/") {
+				if filename := filepath.Base(filepath.FromSlash(params["filename"])); filename != "" {
+					return filename
+				}
+			}
+		}
+	}
+	return ""
+}
+
+func downloadSource(output io.Writer, stdout io.Writer, srcURL, dest string) (remote builder.Source, p string, err error) {
 	u, err := url.Parse(srcURL)
 	if err != nil {
 		return
 	}
-	filename := filepath.Base(filepath.FromSlash(u.Path)) // Ensure in platform semantics
+	filename := getTargetFilename(u.Path)
 	if filename == "" {
-		err = errors.Errorf("cannot determine filename from url: %s", u)
-		return
+		filename = getTargetFilename(dest)
```

Comment:
Nit: please move this down next  to the other `filename` setting code, that way it's all together.

I think this (and the lines below that get the filename from response) could also be in a separate function `getFilenameForDownload(url url.Url, resp http.Response) string` which would mean that you could test the full logic together.